wellcomeml/ml/keras_utils.py

```diff
@@ -15,3 +18,95 @@ def on_epoch_end(self, epoch, logs):
         r = round(recall_score(Y_val, Y_pred, average='micro'), 4)
         print(f" - val metrics: P {p:.4f} R {r:.4f} F {f1:.4f}")
         return
+
+
+class CategoricalMetrics(tf.keras.metrics.Metric):
+    def __init__(self, metric='precision', from_logits=True,
+                 threshold=0.5,
+                 binary='true', pos=1, **kwargs):
+        """
+        Categorical metrics
+
+        Args:
+            metric: 'precision', 'recall' or 'f1'
+            from_logits: Similar to keras 'from_logits'. If True, calculates
+                        a softmax activation on input
+            threshold: a threshold to calculate predictions
+            binary: whether input is binary or not (in which case does average)
+            pos: the positive label (between [0, n_classes])
+            **kwargs:
+        """
+        self.metric = metric
+        name = f"categorical_{metric}"
+        super().__init__(name=name, **kwargs)
+        self.threshold = threshold
+        self.from_logits = from_logits
+        self.binary = binary
+        self.pos = pos
+        self.value = self.add_weight(name='tp', initializer='zeros')
+
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        if self.from_logits:
+            y_pred = tf.keras.activations.softmax(y_pred)
+
+        greater_than_threshold = tf.cast(y_pred[:, self.pos:] > self.threshold,
+                                         'bool')
+        positive = tf.cast(y_true, 'int32') == self.pos
+
+        # Epsilon added to denominators to avoid division by zero
+        eps = tf.keras.backend.epsilon()
+
+        tp = tf.reduce_sum(
+            tf.cast(
+                tf.logical_and(greater_than_threshold, positive),
+                'float32'
+            )
+        )
+
+        if self.metric in ['precision', 'f1_score']:
+            tp_plus_fp = tf.reduce_sum(
```

Comment:
would it not be more general to calculate tp, fp, fn and calculate all metrics accordingly

Reply:
That makes sense, the idea is to add more metrics in the future, so definitely worth thinking better ways to do this. For now the different `if` statements save a bit of calculations...